internal/metrics/metrics.go

```diff
@@ -1,17 +1,224 @@
 )
 
-// Returns an HTTP handler for Prometheus metrics.
-func NewPrometheusMetricsHandler() http.Handler {
-	reg := prometheus.NewRegistry()
-	reg.MustRegister(collectors.NewGoCollector(), collectors.NewProcessCollector(collectors.ProcessCollectorOpts{}))
-	prometheusHandler := promhttp.HandlerFor(reg, promhttp.HandlerOpts{})
-	return prometheusHandler
+type PrometheusMetrics struct {
+	http.Handler
+	registry                       *prometheus.Registry
+	CertificateRequests            prometheus.Gauge
+	OutstandingCertificateRequests prometheus.Gauge
+	Certificates                   prometheus.Gauge
+	CertificatesExpiringIn1Day     prometheus.Gauge
+	CertificatesExpiringIn7Days    prometheus.Gauge
+	CertificatesExpiringIn30Days   prometheus.Gauge
+	CertificatesExpiringIn90Days   prometheus.Gauge
+	ExpiredCertificates            prometheus.Gauge
+
+	RequestsTotal    prometheus.CounterVec
+	RequestsDuration prometheus.HistogramVec
+}
+
+// NewMetricsSubsystem returns the metrics endpoint HTTP handler and the Prometheus metrics for the server and middleware.
+func NewMetricsSubsystem(db *certdb.CertificateRequestsRepository) *PrometheusMetrics {
+	metricsBackend := newPrometheusMetrics()
+	metricsBackend.Handler = promhttp.HandlerFor(metricsBackend.registry, promhttp.HandlerOpts{})
+	ticker := time.NewTicker(120 * time.Second)
+	go func() {
+		for ; ; <-ticker.C {
+			csrs, err := db.RetrieveAll()
+			if err != nil {
+				log.Println(errors.Join(errors.New("error generating metrics repository: "), err))
+				panic(1)
+			}
+			metricsBackend.GenerateMetrics(csrs)
+		}
+	}()
+	return metricsBackend
+}
+
+// newPrometheusMetrics reads the status of the database, calculates all of the values of the metrics,
+// registers these metrics to the prometheus registry, and returns the registry and the metrics.
+// The registry and metrics can be modified from this struct from anywhere in the codebase.
+func newPrometheusMetrics() *PrometheusMetrics {
+	m := &PrometheusMetrics{
+		registry:                       prometheus.NewRegistry(),
+		CertificateRequests:            certificateRequestsMetric(),
+		OutstandingCertificateRequests: outstandingCertificateRequestsMetric(),
+		Certificates:                   certificatesMetric(),
+		ExpiredCertificates:            expiredCertificatesMetric(),
+		CertificatesExpiringIn1Day:     certificatesExpiringIn1DayMetric(),
+		CertificatesExpiringIn7Days:    certificatesExpiringIn7DaysMetric(),
+		CertificatesExpiringIn30Days:   certificatesExpiringIn30DaysMetric(),
+		CertificatesExpiringIn90Days:   certificatesExpiringIn90DaysMetric(),
+
+		RequestsTotal:    requestsTotalMetric(),
+		RequestsDuration: requestDurationMetric(),
+	}
+	m.registry.MustRegister(m.CertificateRequests)
+	m.registry.MustRegister(m.OutstandingCertificateRequests)
+	m.registry.MustRegister(m.Certificates)
+	m.registry.MustRegister(m.ExpiredCertificates)
+	m.registry.MustRegister(m.CertificatesExpiringIn1Day)
```

Comment:
Why not do something like 
`m.registry.MustRegister(certificatesExpiringIn1DayMetric())` and get rid of these variables like `CertificatesExpiringIn1Day` in `PrometheusMetrics`?

Reply:
We save the metrics objects because we need to access them whenever we want to update the values in them. In this case we don't use them directly because we aren't using it outside of this file, but they are public so that if we have a better way to collect or look at any metrics in the future, we don't need to do any complex logic to inspect it. For example, if we want to show one of the metrics in the front end it would be trivial to just look at this object within a handler.

Reply:
Alright, thanks for explaining.